### PR TITLE
ARROW-17104: [C++] Workaround protobuf ABI issues

### PR DESCRIPTION
**Pull request by pitrou**
The protocol buffers ABI unfortunately depends on whether NDEBUG is enabled or not.
When installing gRPC using a package manager (Homebrew, conda...), most often it is built with NDEBUG enabled.
Some hackery is then needed to avoid missing symbols on Arrow debug builds...

Should also fix ARROW-16520.

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@lidavidm @raulcd This at least works for me with the conda-forge libprotobuf package.

File: cpp/src/arrow/util/protobuf_abi_fixup_internal.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+// TODO explain this
```
Comment (lidavidm) [inline cpp/src/arrow/util/protobuf_abi_fixup_internal.h:20]: Would a link to the various JIRAs/Protobuf GitHub issue suffice?

**Review by lidavidm — APPROVED** [review]

**Comment by raulcd** [comment]
Do we require a higher protobuf version for substrait with this change? I think the failures on Ubuntu 18.04 are due to version `3.0.0` being used. 
`Found Protobuf: /usr/lib/x86_64-linux-gnu/libprotobuf.so;-pthread (found suitable version "3.0.0", minimum required is "3.0.0")`
From what I can see `metadata_lite.h` was added on >=3.3.x
Not present on 3.2 branch:
https://github.com/protocolbuffers/protobuf/tree/3.2.x/src/google/protobuf
present on 3.3.:
https://github.com/protocolbuffers/protobuf/tree/3.3.x/src/google/protobuf

**Comment by pitrou** [comment]
@raulcd No, this should hopefully be fixed using a compile-time check.

File: cpp/src/arrow/compute/exec/exec_plan.h [hunk 2]
```diff
@@ -22,6 +22,10 @@
 #include <string>
 #include <vector>
 
+// FIXME there are too many inclusions here; besides, arrow/compute/exec/util.h
+// includes even more, including arrow/util/logging.h which is forbidden
+// in public headers.
```
Comment (lidavidm) [inline cpp/src/arrow/compute/exec/exec_plan.h:27]: Filed https://issues.apache.org/jira/browse/ARROW-17109

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
I usually SSH into the Ursa Mac machines to debug this sort of issue. I could take a look but it might be a few days.

**Comment by pitrou** [comment]
@paleolimbot Can you help me debug this R CI failure? I cannot even reproduce locally (build succeeds here with Docker):
https://github.com/apache/arrow/runs/7393839585?check_suite_focus=true#step:7:4753

**Comment by pitrou** [comment]
Well, `arrow/engine/substrait/api.h` includes `arrow/engine/substrait/serde.h`.

**Comment by pitrou** [comment]
I'm a trying a cache-less build...

**Comment by pitrou** [comment]
Ok, I'm relieved that the issue isn't due to caching at least.

File: cpp/src/arrow/compute/exec/exec_plan.h [hunk 3]
```diff
@@ -22,6 +22,10 @@
 #include <string>
 #include <vector>
 
+// FIXME there are too many inclusions here; besides, arrow/compute/exec/util.h
```
Comment (nealrichardson) [inline cpp/src/arrow/compute/exec/exec_plan.h:25]:
```suggestion
// TODO(ARROW-17109): there are too many inclusions here; besides, arrow/compute/exec/util.h
```

**Comment by pitrou** [comment]
Uh, so including `arrow/engine/substrait/serde.h` explicitly actually fixed the R build failures. I'm puzzled.

**Comment by kou** [comment]
The link error is occurred by our bundled OpenTelemetry C++ not `libarrow*.dylib`.

It will be fixed by using OpenTelemetry C++ installed by Homebrew (but we need https://github.com/Homebrew/homebrew-core/pull/106115 )

```diff
diff --git a/cpp/Brewfile b/cpp/Brewfile
index 9cffd8e3a8..8d709887e6 100644
--- a/cpp/Brewfile
+++ b/cpp/Brewfile
@@ -33,6 +33,7 @@ brew "lz4"
 brew "ninja"
 brew "numpy"
 brew "openssl@1.1"
+brew "opentelemetry-cpp"
 brew "protobuf"
 brew "python"
 brew "rapidjson"
```

or building bundled OpenTelemetry C++ with `CMAKE_BUILD_TYPE=Release`

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 5d1da18b77..8c969cff94 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -4387,10 +4387,16 @@ macro(build_opentelemetry)
     list(APPEND OPENTELEMETRY_LIBRARIES opentelemetry-cpp::${_OPENTELEMETRY_LIB})
   endforeach()
 
+  if(PROTOBUF_VENDORED)
+    set(OPENTELEMETRY_CMAKE_BUILD_TYPE ${CMAKE_BUILD_TYPE})
+  else()
+    set(OPENTELEMETRY_CMAKE_BUILD_TYPE Release)
+  endif()
   set(OPENTELEMETRY_CMAKE_ARGS
       ${EP_COMMON_TOOLCHAIN}
       "-DCMAKE_INSTALL_PREFIX=${OPENTELEMETRY_PREFIX}"
-      "-DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}"
+      "-DCMAKE_BUILD_TYPE=${OPENTELEMETRY_CMAKE_BUILD_TYPE}"
       -DCMAKE_INSTALL_LIBDIR=lib
       "-DCMAKE_CXX_FLAGS=${EP_CXX_FLAGS}"
       -DBUILD_TESTING=OFF
```

**Comment by pitrou** [comment]
Thanks @kou ! Hopefully that will fix the macOS issue, though I'm afraid the Python one might be different.

**Comment by pitrou** [comment]
By the way, @kou, do you know what the issue is with the macOS GLib build? The failure seems unrelated:
```
Program glib-mkenums mkenums found: NO
c_glib/arrow-glib/meson.build:216:0: ERROR: Program 'glib-mkenums mkenums' not found or not executable
```

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g python

**Comment by pitrou** [comment]
Ok, the remaining problem is the macOS Python symbol lookup error which, for some reason, wasn't solved by this PR.

**Comment by pitrou** [comment]
Ok, my latest attempt failed again.

**Comment by pitrou** [comment]
Well, finally it seems the issue may be fixed in the next upstream release:
https://github.com/protocolbuffers/protobuf/pull/10271

**Comment by kou** [comment]
I've fixed GLib and Python failures.

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g python

**Comment by kou** [comment]
https://app.travis-ci.com/github/apache/arrow/jobs/577173431#L2887

```text
In file included from /arrow/cpp/src/arrow/flight/protocol_internal.h:22,
                 from /arrow/cpp/src/arrow/flight/serialization_internal.h:22,
                 from /arrow/cpp/src/arrow/flight/client.cc:39,
                 from src/arrow/flight/CMakeFiles/arrow_flight_objlib.dir/Unity/unity_0_cxx.cxx:4:
/arrow/cpp/src/arrow/util/protobuf_abi_fixup_internal.h:39:2: error: #error "Protobuf already included, fixup can't be applied!"
   39 | #error "Protobuf already included, fixup can't be applied!"
      |  ^~~~~
```

It seems that we need a wrapper for `Flight.pb.cc` like `cpp/src/arrow/flight/transport/grpc/protocol_grpc_internal.cc`.

**Comment by kou** [comment]
I think that we need this because other packaging system may ship Protobuf that has this problem.
If you don't want to complete this, can I take over this?

**Comment by pitrou** [comment]
> I think that we need this because other packaging system may ship Protobuf that has this problem.

Which packaging systems ship the offending protobuf version?
This is mainly for us and our CI. If you compile Arrow in release mode, there is no problem. Third-party users are unlikely to compile Arrow in debug mode...

> If you don't want to complete this, can I take over this?

@kou You can. But the fix is extremely brittle, so I'm in favour of not doing it since it's fixed upstream.

**Comment by kou** [comment]
> This is mainly for us and our CI. If you compile Arrow in release mode, there is no problem.

Ah, I see. OK. We can close this.

**Comment by raulcd** [comment]
Should we still merge https://github.com/apache/arrow/pull/13634/commits/9e10f6c3399d83ebce5af551561fa3a16da9cd5e independently? I think is required to fix the:
```
Program glib-mkenums mkenums found: NO
c_glib/arrow-glib/meson.build:216:0: ERROR: Program 'glib-mkenums mkenums' not found or not executable
```
Which happens on master?

**Comment by pitrou** [comment]
Right, and the added `cpp/src/arrow/public_api_test.cc` test (together with the necessary source code fixes) would be nice to get in as well. So let's keep this PR open until they're addressed by another PR?

**Comment by pitrou** [comment]
I opened https://github.com/apache/arrow/pull/13965 for the public API test improvement.

**Pull request closed without merging**